File: Social/Safety/SafetyManager.swift
//
//  SafetyManager.swift
//  Social
//
//  Bloqueo, denuncia, modo invisible y verificación de cuenta. Debe ser
//  accesible desde cualquier pantalla (principio de producto no negociable),
//  no solo desde este archivo — la app inyecta este manager como
//  @EnvironmentObject en la raíz para que cualquier vista pueda llamarlo.
//

import Foundation
import SwiftUI

@MainActor
final class SafetyManager: ObservableObject {

    @Published var errorMessage: String?

    /// Activa/desactiva el modo invisible del usuario actual (profiles.is_invisible).
    /// Debe poder activarse en un toque desde cualquier pantalla.
    func setInvisible(_ invisible: Bool, userID: UUID) async {
        do {
            try await SupabaseManager.shared.client
                .from("profiles")
                .update(["is_invisible": invisible])
                .eq("id", value: userID)
                .execute()
        } catch {
            errorMessage = "No se pudo cambiar el modo invisible."
        }
    }

    /// Bloquea a otro usuario: deja de ser descubrible y no puede escribir ni enviar socials.
    /// Se modela reutilizando `socials` con status "declined" + una tabla ligera de bloqueos.
    func block(userID: UUID, blockedID: UUID) async {
        struct BlockRow: Encodable {
            let blocker_id: UUID
            let blocked_id: UUID
        }
        do {
            try await SupabaseManager.shared.client
                .from("blocks")
                .insert(BlockRow(blocker_id: userID, blocked_id: blockedID))
                .execute()
            // Hallazgo real, mismo criterio ya aplicado en la versión
            // Kotlin equivalente: bloquear (una métrica real de
            // confianza y seguridad) tampoco se registraba.
            AnalyticsManager.track("block_created")
        } catch {
            errorMessage = "No se pudo bloquear a este usuario."
        }
    }

    /// Restringir una cuenta real, comparado con Instagram -- deliberadamente
    /// MÁS SUAVE que bloquear (arriba): sus comentarios en TUS publicaciones/
    /// reels dejan de verse para todos los demás (siguen viéndose con
    /// normalidad para quien los escribió, que nunca se entera de nada). A
    /// diferencia de block(), nunca se avisa ni se nota nada del lado de la
    /// persona restringida -- ver 0093_restrict_account.sql. Equivalente de
    /// SafetyManager.kt.restrict().
    func restrict(userID: UUID, restrictedID: UUID) async {
        struct RestrictRow: Encodable {
            let restricter_id: UUID
            let restricted_id: UUID
        }
        do {
            try await SupabaseManager.shared.client
                .from("restricts")
                .insert(RestrictRow(restricter_id: userID, restricted_id: restrictedID))
                .execute()
        } catch {
            errorMessage = "No se pudo restringir a este usuario."
        }
    }

    func unrestrict(userID: UUID, restrictedID: UUID) async {
        do {
            try await SupabaseManager.shared.client
                .from("restricts")
                .delete()
                .eq("restricter_id", value: userID)
                .eq("restricted_id", value: restrictedID)
                .execute()
        } catch {
            errorMessage = "No se pudo deshacer la restricción."
        }
    }

    /// Envía una denuncia. Se revisa manualmente por moderación (fuera del alcance de la app cliente).
    func report(
        reporterID: UUID, reportedID: UUID, reason: String, details: String?,
        postID: UUID? = nil, commentID: UUID? = nil, messageID: UUID? = nil, groupMessageID: UUID? = nil
    ) async {
        // Mismo límite real que reports_details_length
        // (0024_more_text_length_limits.sql) — "details" es el único
        // campo libre de este formulario, ya construido en la versión
        // Kotlin equivalente.
        guard (details?.count ?? 0) <= 1000 else {
            errorMessage = "Los detalles no pueden tener más de 1000 caracteres."
            return
        }
        struct ReportRow: Encodable {
            let reporter_id: UUID
            let reported_id: UUID
            let reason: String
            let details: String?
            // Hallazgo real, comparado con Instagram/TikTok/Facebook: antes
            // el único rastro de "esta denuncia es sobre ESTE post/
            // comentario" era un texto libre y editable metido a mano en
            // `details` -- referencia real ahora
            // (0045_reports_content_reference.sql).
            let post_id: UUID?
            let comment_id: UUID?
            // Hallazgo real, comparado con Instagram/WhatsApp/Messenger:
            // mismo hueco exacto que post_id/comment_id pero en un chat --
            // denunciar desde un chat solo apuntaba al perfil de la otra
            // persona, sin ningún rastro de QUÉ mensaje concreto motivó la
            // denuncia (0048_reports_message_reference.sql).
            let message_id: UUID?
            // Mismo hueco exacto que message_id pero en un chat de grupo --
            // ver 0067_reports_group_message_reference.sql.
            let group_message_id: UUID?
        }
        do {
            try await SupabaseManager.shared.client
                .from("reports")
                .insert(ReportRow(
                    reporter_id: reporterID, reported_id: reportedID, reason: reason, details: details,
                    post_id: postID, comment_id: commentID, message_id: messageID, group_message_id: groupMessageID
                ))
                .execute()
            // Hallazgo real, mismo criterio ya aplicado en la versión
            // Kotlin equivalente: cada acción clave de la app se registra
            // con AnalyticsManager (duel_completed, tab_view, app_open...)
            // salvo denunciar.
            AnalyticsManager.track("report_submitted")
        } catch {
            errorMessage = "No se pudo enviar la denuncia."
        }
    }

    /// Verificación de cuenta: compara una selfie en vivo contra el avatar guardado.
    /// La comparación real de similitud debe hacerla un servicio backend (no el cliente),
    /// para no exponer el modelo de comparación ni permitir manipulación local del resultado.
    func requestVerification(userID: UUID, liveSelfie: UIImage) async -> Bool {
        // Placeholder: en producción, sube la selfie a una Edge Function que la
        // compara contra avatar_url y actualiza profiles.is_verified server-side.
        errorMessage = "La verificación de cuenta requiere el servicio backend de comparación (pendiente de implementar en Supabase Edge Functions)."
        return false
    }
}

File: Android/app/src/main/java/com/social/app/safety/SafetyManager.kt
package com.social.app.safety

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.social.app.backend.SupabaseManager
import io.github.jan.supabase.gotrue.auth
import io.github.jan.supabase.postgrest.from
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Bloqueo, denuncia y modo invisible — equivalente Kotlin de
 * SafetyManager.swift. Mismas tablas (`blocks`, `reports`, `profiles`), mismo
 * RLS del backend compartido.
 */
class SafetyManager : ViewModel() {

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    @Serializable
    private data class InvisibleUpdate(@SerialName("is_invisible") val isInvisible: Boolean)

    /** Persiste el modo invisible en el perfil. El efecto real sobre el
     * motor UWB (dejar de anunciarse) se hace por separado, llamando a
     * `SocialProximity.setDiscoverable()` — ver comentario en iOS sobre
     * por qué esto tiene que actuar en dos capas. */
    fun setInvisible(invisible: Boolean, userId: String) {
        viewModelScope.launch {
            try {
                SupabaseManager.client.from("profiles")
                    .update(InvisibleUpdate(invisible)) { filter { eq("id", userId) } }
            } catch (e: Exception) {
                _errorMessage.value = "No se pudo cambiar el modo invisible."
            }
        }
    }

    @Serializable
    private data class BlockRow(
        @SerialName("blocker_id") val blockerId: String,
        @SerialName("blocked_id") val blockedId: String
    )

    fun block(userId: String, blockedId: String) {
        viewModelScope.launch {
            try {
                SupabaseManager.client.from("blocks").insert(BlockRow(userId, blockedId))
                // Hallazgo real, misma auditoría de AnalyticsManager de
                // la pasada anterior: bloquear (una métrica real de
                // confianza y seguridad) tampoco se registraba.
                com.social.app.backend.AnalyticsManager.track("block_created")
            } catch (e: Exception) {
                _errorMessage.value = "No se pudo bloquear a este usuario."
            }
        }
    }

    @Serializable
    private data class ReportRow(
        @SerialName("reporter_id") val reporterId: String,
        @SerialName("reported_id") val reportedId: String,
        val reason: String,
        val details: String?,
        // Hallazgo real, comparado con Instagram/TikTok/Facebook: antes el
        // único rastro de "esta denuncia es sobre ESTE post/comentario"
        // era un texto libre y editable ("Publicación {id}" metido a mano
        // en `details`) -- un admin no tenía forma real de ver el
        // contenido denunciado. Referencia real (0045_reports_content_reference.sql).
        @SerialName("post_id") val postId: String? = null,
        @SerialName("comment_id") val commentId: String? = null
    )

    fun report(
        reporterId: String,
        reportedId: String,
        reason: String,
        details: String?,
        postId: String? = null,
        commentId: String? = null
    ) {
        // Mismo límite real que reports_details_length
        // (0024_more_text_length_limits.sql) — "details" es el único
        // campo libre de este formulario ("reason" es una de las
        // REASONS fijas de ReportSheet.kt, no texto libre).
        if ((details?.length ?: 0) > 1000) {
            _errorMessage.value = "Los detalles no pueden tener más de 1000 caracteres."
            return
        }
        viewModelScope.launch {
            try {
                SupabaseManager.client.from("reports").insert(ReportRow(reporterId, reportedId, reason, details, postId, commentId))
                // Hallazgo real: cada acción clave de la app se registra
                // con AnalyticsManager (duel_completed, tab_view,
                // app_open...) salvo denunciar — el propio equipo de
                // SOCIAL no tendría forma de saber si esta función se usa
                // de verdad sin entrar a la tabla `reports` a mano.
                com.social.app.backend.AnalyticsManager.track("report_submitted")
            } catch (e: Exception) {
                _errorMessage.value = "No se pudo enviar la denuncia."
            }
        }
    }

    /**
     * Verificación de cuenta: compara una selfie en vivo contra el avatar
     * guardado. Igual que en `SafetyManager.swift`, la comparación real de
     * similitud debe hacerla un servicio backend (Edge Function), nunca el
     * cliente — para no exponer el modelo de comparación ni permitir
     * manipulación local del resultado. Placeholder honesto, no una
     * implementación simulada: el resultado real vendría de una Edge
     * Function equivalente a `duel-ai` que reciba la selfie, la compare
     * contra `avatar_url` y actualice `profiles.is_verified` server-side.
     */
    suspend fun requestVerification(): Boolean {
        _errorMessage.value = "La verificación de cuenta requiere el servicio backend de comparación (pendiente de implementar en Supabase Edge Functions)."
        return false
    }
}
